File: VariAuto/Codes/SGD.m
classdef SGD < Trainer

    properties (Access = protected)
       learningRate  
    end

    properties (Access = private)
       thetaLowest
       fvStop
       lSearchtype
       MaxEpochs
       maxFunEvals
       optTolerance
       earlyStop
       timeStop
       plotter
       svepoch
       fplot
    end

    methods (Access = public)

        function obj = SGD(s)
            obj.init(s)
            obj.plotter = s.plotter;
            obj.learningRate = s.learningRate;
            obj.maxFunEvals  = 5000;
            obj.optTolerance = 10^(-8);
            obj.timeStop    = Inf([1,1]);
            obj.fvStop      = -Inf;%10^(-4);
            obj.nPlot       = 1;
            obj.MaxEpochs   = 1000;
            obj.earlyStop   = obj.MaxEpochs;
            obj.svepoch     = 0;
            obj.lSearchtype  = 'static';
        end
        
        function train(obj)
           tic
           x0  = obj.designVariable.thetavec;
           F = @(theta,order,i) obj.objectiveFunction.computeCost(theta,order,i); 
           obj.optimize(F,x0);
           toc
        end

        function plotCostFunc(obj)
            figure(3);
            % epoch = 1:obj.MaxEpochs;
            epoch = 1:length(obj.fplot);
            % plot(epoch,obj.fplot,'-o');
            plot(epoch,obj.fplot,'LineWidth',1.8);
            xlabel('Epochs')
            ylabel('Function Values')
            title('Cost Function')
            xlim([1,inf])
        end

    end
    
    methods(Access = private)  
        
        function optimize(obj,F,th0)
            [nD, nB, ~] = obj.objectiveFunction.getBatchSize();
            epsilon0      = obj.learningRate;
            epoch         =  1;iter = -1; funcount =  0; fv = 1;
            alarm         =  0; gnorm = 1; min_testError = 1;
            criteria(1)   = epoch <= obj.MaxEpochs; 
            criteria(2)   = alarm < obj.earlyStop; 
            criteria(3)   = gnorm > obj.optTolerance;
            criteria(4)   = toc < obj.timeStop;
            criteria(5)   = fv > obj.fvStop;
            while all(criteria == 1)

                if nB == 1 || nB == 0
                    order = 1:nD;
                    nB = 1;
                else
                    order = randperm(nD,nD);
                end
                for i = 1:nB
                    if iter == -1
                        th      = th0;
                        epsilon = epsilon0;
                        state   = 'init';   
                    else
                        state   = 'iter';
                    end
                    [f,grad]              = F(th,order, i);  
                    [epsilon,th,funcount] = obj.lineSearch(th,grad,F,f,epsilon,epsilon0,funcount,order,i);                
                    gnorm                 = norm(grad,2);
                    opt.epsilon           = epsilon*gnorm; 
                    opt.gnorm             = gnorm;
                    funcount              = funcount + 1;
                    iter                  = iter + 1;
                    obj.displayIter(epoch,iter,funcount,th,f,opt,state);
                end
                [alarm,min_testError] = obj.validateES(alarm,min_testError);
                epoch = epoch + 1;
                criteria(1)   = epoch <= obj.MaxEpochs; 
                criteria(2)   = alarm < obj.earlyStop; 
                criteria(3)   = gnorm > obj.optTolerance;
                criteria(4)   = toc < obj.timeStop; 
                criteria(5)   = f > obj.fvStop;
            end
            if criteria(1) == 0
                fprintf('Minimization terminated, maximum number of epochs reached %d\n',epoch)
            elseif criteria(2) == 0
                fprintf('Minimization terminated, validation set did not decrease in %d epochs\n',obj.earlyStop)
            elseif criteria(3) == 0
                fprintf('Minimization terminated, reached the optimalty tolerance of %f\n',obj.optTolerance)
            elseif criteria(4) == 0
                fprintf('Minimization terminated, reached the limit time of %f\n',obj.timeStop)
            elseif criteria(5) == 0
                fprintf('Minimization terminated, reached the target function value of %f\n',obj.fvStop)
            else
                fprintf('The operation terminated excepcionally\n')
            end
            %F(th,Xb,Yb);
            %F(th,order, i);
            %th
        end

        function [e,x,funcount] = lineSearch(obj,x,grad,F,fOld,e,e0,funcount,order,i)
            type = obj.lSearchtype;
            switch type
                case 'static'
                    xnew = obj.step(x,e,grad);
                case 'decay'
                    tau = 50;
                    xnew = obj.step(x,e,grad);
                    e = e - 0.99*e0*30/tau;
                case 'dynamic'
                    f = fOld;
                    xnew = x;
                    while f >= 1.001*(fOld - e*(grad*grad'))
                        xnew = obj.step(x,e,grad);
                        [f,~] = F(xnew,order,i);
                        e = e/2;
                        funcount = funcount + 1;
                    end
                    e = 5*e;                
                case 'fminbnd'
                    xnew = @(e1) obj.step(x,e1,grad);
                    f = @(e1) F(xnew(e1),order,i);
                    [e,~] = fminbnd(f,e/10,e*10);
                    xnew = xnew(e);
            end
            x = xnew;
        end 

        function [alarm,min_testError] = validateES(obj,alarm,min_testError)
            [Xtest, Ytest] = obj.objectiveFunction.getTestData();
            [~,y_pred]   = max(obj.objectiveFunction.getOutput(Xtest),[],2);
            [~,y_target] = max(Ytest,[],2);
            testError    = mean(y_pred ~= y_target);
            if testError < min_testError
                obj.thetaLowest = obj.designVariable.thetavec;
                min_testError = testError;
                alarm = 0;
            elseif testError == min_testError
                alarm = alarm + 0.5;
            else
                alarm = alarm + 1;
            end
        end

        function displayIter(obj,epoch,iter,funcount,x,f,opt,state)
            [nD, ~, batchSize] = obj.objectiveFunction.getBatchSize();
            obj.printValues(epoch,funcount,opt,f,iter)
            if obj.isDisplayed == true
                if iter*batchSize==(epoch-1)*nD || iter == -1
                    obj.storeValues(x,f,state,opt);
                    obj.plotMinimization(epoch);
                end
            end
        end  

        function printValues(obj,epoch,funcount,opt,f,iter)
            formatstr = '%5.0f    %5.0f       %5.0f    %13.6g  %13.6g   %12.3g\n';
            if mod(iter,20) == 0
                fprintf(['                                                        First-order \n', ...
                    'Epoch Iteration  Func-count       f(x)        Step-size       optimality\n']);
            end
            fprintf(formatstr,epoch,iter,funcount,f,opt.epsilon,opt.gnorm);
            
            if epoch ~= obj.svepoch
                obj.svepoch = epoch;
                obj.fplot(1,epoch) = f;
            end

        end

    end
    
    methods (Access = protected)
        function x = step(obj,x,e,grad)
            x = x - e*grad;
        end
    end
end

File: Tutorials/OpenArmsTutorial/Gridap/SGD.jl
module SGD

export SGDStruct, init_SGD, compute

using ..Trainer
using ..CostNN
using Dates
using Optim
using Plots
using Printf
using LinearAlgebra

"""
    SGDStruct

Immutable structure storing SGD hyperparameters, trainer, and state history.
"""
struct SGDStruct
    trainer::Trainer.TrainerStruct
    fvStop::Float64
    l_search_type::String
    max_epochs::Int
    max_fun_evals::Int
    opt_tolerance::Float64
    early_stop::Int
    time_stop::Float64
    learning_rate::Float64
    svepoch::Int
    fplot::Vector{Float64}
    elapsed_time::Float64
end

"""
    init_SGD(params)

Initializes SGDStruct from parameter dictionary.
"""
function init_SGD(params::Dict{String, Any})
    trainer = Trainer.init_Trainer(params)

    return SGDStruct(
        trainer,
        1e-4,                       # fvStop
        "static",                   # l_search_type
        params["maxEpochs"],
        5000,                       # max_fun_evals
        1e-8,                       # opt_tolerance
        params["maxEpochs"],
        Inf,                        # time_stop
        params["learningRate"],
        0,                          # svepoch
        fill(0.0, params["maxEpochs"]),  # fplot
        0.0                         # elapsed_time
    )
end

function compute(sgd::SGDStruct, θ::Vector{Float64})
    start_time = time()

    kpi = (epoch=1, alarm=0.0, gnorm=1.0, cost=1.0)

    sgd, _ = _optimize(sgd, θ, kpi, start_time)

    elapsed = time() - start_time
    sgd = update_elapsed_time(sgd, elapsed)

    println("Elapsed time: $(round(elapsed, digits=4)) seconds")
    return sgd
end

function _optimize(sgd::SGDStruct, θ::Vector{Float64}, kpi, start_time::Float64)
    ε = sgd.learning_rate
    min_test_error = 1.0
    iter = -1
    funcount = 0

    while !is_criteria_met(sgd, kpi, time()-start_time)
        new_epoch = true
        move_batch = true
        is_batch_depleted = false
        objFunc = sgd.trainer.objectiveFunction

        while !is_batch_depleted || new_epoch
            # Compute stochastic function and gradient
            f, grad, is_batch_depleted, _= compute_stochastic_function_and_gradient(objFunc, θ, move_batch)

            # Line search
            θ_new, ε, funcount_inc = line_search(sgd, θ, grad, f, ε)
            θ = θ_new

            # Update KPI
            gnorm = norm(grad)
            kpi = (epoch=kpi.epoch, alarm=kpi.alarm, gnorm=gnorm, cost=f)

            # Update fplot and svepoch
            sgd = update_fplot_and_svepoch(sgd, kpi.epoch, f)

            # Display iteration info
            optinfo = Trainer.opt_info(ε * gnorm, gnorm)
            sgd = display_iter(sgd, iter, funcount, θ, kpi, optinfo)

            funcount += 1   # For compute_stochastic_function_and_gradient call
            funcount += funcount_inc   # For line search internal calls
            
            iter += 1
            new_epoch = false
        end

        # Epoch increment
        kpi = (epoch=kpi.epoch+1, alarm=kpi.alarm, gnorm=kpi.gnorm, cost=kpi.cost)

        # Early stopping validation
        alarm, min_test_error = CostNN.validate_ES(objFunc, kpi.alarm, min_test_error, θ)
        kpi = (epoch=kpi.epoch, alarm=alarm, gnorm=kpi.gnorm, cost=kpi.cost)
    end

    return sgd, θ
end

function compute_stochastic_function_and_gradient(objFunc, θ::Vector{Float64}, move_batch::Bool)
    objFunc = CostNN.set_batch_mover(objFunc, move_batch)
    jV, djV, is_batch_depleted, Jc = CostNN.compute_stochastic_function_and_gradient(objFunc, θ)
    return jV, djV, is_batch_depleted, Jc
end

"""
    line_search(sgd, θ, grad, fOld, ε)

Returns updated θ, learning rate, and funcount increment.
"""
function line_search(sgd::SGDStruct, θ::Vector{Float64}, grad::Vector{Float64}, fOld::Float64, ε::Float64)
    type = sgd.l_search_type
    θ_new = copy(θ)
    funcount_inc = 0

    if type == "static"
        θ_new .= step(θ, ε, grad)

    elseif type == "decay"
        θ_new .= step(θ, ε, grad)
        ε *= (1 - 1e-3)

    elseif type == "dynamic"
        f = fOld
        while f >= 1.001 * (fOld - ε * dot(grad, grad))
            θ_new .= step(θ, ε, grad)
            f, _, _, _ = compute_stochastic_function_and_gradient(sgd.trainer.objectiveFunction, θ_new, false)
            ε /= 2
            funcount_inc += 1
        end
        ε *= 5

    elseif type == "fminbnd"
        F = θ -> compute_stochastic_function_and_gradient(sgd.trainer.objectiveFunction, θ, false)[1]
        f_e(e1) = F(step(θ, e1, grad))
        result = Optim.optimize(f_e, ε/10, ε*10)
        ε = Optim.minimizer(result)
        θ_new .= step(θ, ε, grad)

    else
        @warn "Unknown line search type: $type. Defaulting to static step."
        θ_new .= step(θ, ε, grad)
    end

    return θ_new, ε, funcount_inc
end

"""
    is_criteria_met(sgd, kpi, elapsed)

Checks all stopping criteria and prints termination message if met.
"""
function is_criteria_met(sgd::SGDStruct, kpi, elapsed::Float64)
    criteria = [
        kpi.epoch <= sgd.max_epochs,
        kpi.alarm < sgd.early_stop,
        kpi.gnorm > sgd.opt_tolerance,
        elapsed < sgd.time_stop,
        kpi.cost > sgd.fvStop
    ]

    failedIdx = findfirst(!, criteria)
    if isnothing(failedIdx)
        return false
    else
        messages = [
            "Minimization terminated: maximum number of epochs reached ($(kpi.epoch))",
            "Minimization terminated: validation set did not decrease in $(sgd.early_stop) epochs",
            "Minimization terminated: reached the optimality tolerance of $(sgd.opt_tolerance))",
            "Minimization terminated: reached the time limit of $(sgd.time_stop)) seconds",
            "Minimization terminated: reached the target function value of $(sgd.fvStop))"
        ]
        println(messages[failedIdx])
        return true
    end
end

"""
    display_iter(sgd, iter, funcount, θ, kpi, optinfo)

Displays iteration info and stores values if needed.
Returns updated SGDStruct.
"""
function display_iter(sgd::SGDStruct, iter::Int, funcount::Int, θ::Vector{Float64}, kpi, optinfo::Trainer.opt_info)
    if iter % 20 == 0
        println("                                                        First-order")
        println("Epoch Iteration  Func-count       f(x)        Step-size       optimality")
    end

    @printf("%5d    %5d       %5d    %13.6g  %13.6g   %12.3g\n",
        kpi.epoch, iter, funcount, kpi.cost, optinfo.ε, optinfo.gnorm)

    if sgd.trainer.isDisplayed && ((kpi.epoch % 25 == 0) || iter == -1)
        updated_trainer = Trainer.store_values(sgd.trainer, θ, kpi.cost, optinfo)
        sgd = update_trainer(sgd, updated_trainer)
    end

    return sgd
end

"""
    update_trainer(sgd, updated_trainer)

Returns SGDStruct with updated trainer.
"""
update_trainer(sgd::SGDStruct, updated_trainer::Trainer.TrainerStruct) = SGDStruct(updated_trainer, sgd.fvStop, sgd.l_search_type,
    sgd.max_epochs, sgd.max_fun_evals, sgd.opt_tolerance, sgd.early_stop, sgd.time_stop,
    sgd.learning_rate, sgd.svepoch, sgd.fplot, sgd.elapsed_time)

"""
    update_fplot_and_svepoch(sgd, epoch, f)

Returns SGDStruct with updated fplot and svepoch.
"""
function update_fplot_and_svepoch(sgd::SGDStruct, epoch::Int, f::Float64)
    new_fplot = copy(sgd.fplot)
    new_fplot[epoch] = f
    return SGDStruct(sgd.trainer, sgd.fvStop, sgd.l_search_type, sgd.max_epochs,
        sgd.max_fun_evals, sgd.opt_tolerance, sgd.early_stop, sgd.time_stop,
        sgd.learning_rate, epoch, new_fplot, sgd.elapsed_time)
end

"""
    update_elapsed_time(sgd, elapsed)

Returns SGDStruct with updated elapsed_time.
"""
update_elapsed_time(sgd::SGDStruct, elapsed::Float64) = SGDStruct(sgd.trainer, sgd.fvStop, sgd.l_search_type,
    sgd.max_epochs, sgd.max_fun_evals, sgd.opt_tolerance, sgd.early_stop, sgd.time_stop,
    sgd.learning_rate, sgd.svepoch, sgd.fplot, elapsed)

"""
    step(θ, ε, grad)

Returns updated θ after gradient step.
"""
step(θ::Vector{Float64}, ε::Float64, grad::Vector{Float64}) = θ .- ε .* grad

end